### ISSUE TITLE
DSRs mess up anti-unification

### ISSUE THREAD

**Issue by nadia-polikarpova** [dcao/babble#10]
Running `smiley` on the following example (without DSRs):

```
(+ 
   (move 1 0 circle)
   (move 2 0 circle)
   (move 3 0 circle)
   (move 4 0 circle)
   (move 5 0 circle)
   (move 6 0 circle)
)
```

produces the expected result:

```
lib l0 =
  λx1 -> move <x1, 0> circle
in
  [l0 1,
    l0 2,
    l0 3,
    l0 4,
    l0 5,
    l0 6]
```

Adding a DSR `"circle" => "(scale 1 circle)"` works well, and even allows to obtain the above result if some of the circles in the input are scaled by 1. 

However, adding a more general rewrite `?x -> (scale 1 ?x)` results in not learning any libraries, and moreover returning an expression that is larger than the original:

```
scale 1 [move <1, 0> circle,
  move <2, 0> circle,
  move <3, 0> circle,
  move <4, 0> circle,
  move <5, 0> circle,
  move <6, 0> circle]
```

There might be two bugs here:
1. AU always returns empty for all pairs of states except the diagonal (might have to do with the e-graph being too loopy?)
2. Extraction also picks the wrong expression (not the original)

**Comment by rosekunkel** [comment]
Even smaller example:
```
(+ circle circle)
```
with the rewrite `?x => (scale 1 ?x)` produces the result
```
scale 1 [circle, circle]
```

**Comment by rosekunkel** [comment]
The example above demonstrates just the extractor bug, this one demonstrates just the library learning bug:
```
(+ (move 2 0 circle)
   (move 1 0 circle))
```
with these two rewrites:
```
1 => (scale 1 1)
2 => (scale 1 2)
```

**Comment by dcao** [comment]
Resolved with a2003cb.
